Implement custom `github:` URL scheme in Go

jandubois opened the pull request:
This is a reimplementation of the bash code in https://github.com/lima-vm/lima/pull/3937#issue-3368758579

AkihiroSuda commented on pkg/limatmpl/locator.go:
```diff
@@ -312,6 +417,15 @@ func TransformCustomURL(ctx context.Context, locator string) (string, error) {
+		return newLocator, nil
+	}
+
 	plugin, err := plugins.Find("url-" + u.Scheme)
```
Any reason not to implement the github locator as a plugin?

jandubois replied:
I had implemented it as a plugin in bash first, but that is not portable to Windows.

I've been thinking of `github:` as been an "internal" scheme, like `template:`, meaning you cannot modify it via plugin, but it will have the same semantics everywhere.

AkihiroSuda commented on pkg/limatmpl/locator.go:
```diff
@@ -296,6 +298,109 @@ func InstNameFromYAMLPath(yamlPath string) (string, error) {
 }
 
+// transformGitHubURL transforms a github: URL to a raw.githubusercontent.com URL.
+// Input format: ORG/REPO[/PATH][@BRANCH]
```
```suggestion
// Input format: ORG/REPO[/PATH][@BRANCH][?digest=ALGO:HEX]
```

jandubois replied:
That seems inconsistent with how we would deal with digests for any other kind of locator. We already had this discussion, and you rejected adding it as a suffix to other URL schemes.

The way to specify a digest is:

```shell
limactl create github:org/repo --digest ALGO:HEX
```

Or in your template:

```yaml
base:
- url: github:ORG/REPO
  digest: ALGO:HEX

provision:
- mode: system
  file:
    url: github:ORG/REPO/FILE
    digest: ALGO:HEX
```

The `--digest` option is not yet implemented, but the `digest` property already exists in the schema, but throws an error when you set the value. It is implemented for base templates, provisioning scripts, and probes.

Adding a digest to the `url` field is confusing, and doesn't work for other kinds of URLs.

AkihiroSuda replied:
> That seems inconsistent with how we would deal with digests for any other kind of locator. We already had this discussion, and you rejected adding it as a suffix to other URL schemes.

I think I rejected extending `https://...` URLs, but I don't think that I have rejected adding `?digest=...` part to `github:` schema (unless the `github:` schema is following some existing spec like purl)

Adding `?digest=...` part might be useful for parameterizing scripts like:
```
limactl create "${TEMPLATE}"
```

HTTPS URLs still have to rely on `--digest` though.

Anyway, this does not need to be implemented in this PR right now

jandubois commented on pkg/limatmpl/locator.go:
```diff
@@ -312,6 +417,15 @@ func TransformCustomURL(ctx context.Context, locator string) (string, error) {
 		return newLocator, nil
 	}
 
+	if u.Scheme == "github" {
```
Sure, let's do this, so we can still tweak the semantics.

jandubois replied:
> Do we want to mark this experimental and print a warning?

I've done this, but not sure I like it because it quickly becomes ugly:

```console
❯ l tmpl url github:jandubois
WARN[0000] GitHub locator "github:jandubois" replaced with "https://raw.githubusercontent.com/jandubois/jandubois/main/templates/demo.yaml" is still EXPERIMENTAL
https://raw.githubusercontent.com/jandubois/jandubois/main/templates/demo.yaml
```

Maybe we can make it non-experimental in Lima 2.1

AkihiroSuda commented:
> github:ORG means github:ORG/ORG

`github:ORG/.github` may make more sense?
Still makes sense to fall-back to `github:ORG/ORG`

> github:ORG//PATH means github:ORG/ORG/PATH

This looks confusing to me. `//` should just be same as `/` ?

jandubois commented:
> > github:ORG means github:ORG/ORG
> 
> `github:ORG/.github` may make more sense?

That doesn't make sense to me, the `.github` repo is for storing GitHub related metadata, not project specific files.

Many projects use the same name for their org and main repo. We would be using `lima/lima` too, had the `lima` org still been available.

> Still makes sense to fall-back to `github:ORG/ORG`

Yet another network roundtrip, and I doubt that `.github` is a common use case.
 
> > github:ORG//PATH means github:ORG/ORG/PATH
> 
> This looks confusing to me. `//` should just be same as `/` ?

It is a direct extension of "an empty REPO means it is the same as the ORG".

I don't see what useful purpose treating extra slashes as redundant would serve.

It is not really different from `file://foo/bar` meaning something completely different from `file:///foo/bar`.

> > github:newlinux to point to github:newlinux/newlinux/templates/newlinux.yaml
> Users probably don't want this file name either, as the file name doesn't explain what this YAML is for.

It creates an instance with the `newlinux` distro or software. The same way we have `rockylinux` and `almalinux` templates.

> Can be easily confused as a Kubernetes manifest.

How so? Why can `github:gardenlinux` be confused with a Kubernetes manifest, but `template:rockylinux` can not?

> Maybe it should be resolved as github:newlinux/newlinux/newlinux.lima.yaml?
> The instance name should be resolved as newlinux omitting the .lima.yaml suffix.

I would prefer not adding another heuristic. Let's keep things simple!

I still think my suggestions are sensible!

AkihiroSuda commented:
>> Can be easily confused as a Kubernetes manifest.
>
> How so? Why can github:gardenlinux be confused with a Kubernetes manifest, but template:rockylinux can not?

What I meant was that `newlinux.yaml` in the repo could be confused as a YAML for Kubernetes or something else that accepts YAML.

AkihiroSuda commented:
An alternative to `REPO.lima.yaml` would be `lima/REPO.yaml` or `.lima/REPO.yaml`.

jandubois commented on pkg/limatmpl/github.go:
```diff
@@ -0,0 +1,196 @@
+	if LooksLikeSymlink(content) {
+		redirectPath := strings.TrimSpace(content)
+		if redirectPath != "" {
+			return redirectPath, nil
```
This is not correct if the `filePath` includes directories because the `redirectPath` will be relative to this directory and not to the repo root.

jandubois replied:
Fixed now:

```console
❯ tree -a -I .git ~/git/jandubois
/Users/jan/git/jandubois
├── .lima.yaml -> templates/demo.yaml
├── docs
│   └── lima.yaml -> ../templates/demo.yaml
└── templates
    └── demo.yaml

3 directories, 3 files

❯ l tmpl url github:jandubois//docs/lima.yaml
WARN[0000] GitHub locator "github:jandubois//docs/lima.yaml" replaced with "https://raw.githubusercontent.com/jandubois/jandubois/main/templates/demo.yaml" is still EXPERIMENTAL
https://raw.githubusercontent.com/jandubois/jandubois/main/templates/demo.yaml
```

jandubois commented:
I've implemented the changes I suggested above. I also created a `jandubois/jandubois` repo for testing:

```console
❯ ls -la ~/git/jandubois
total 0
drwxr-xr-x@  5 jan  staff  160 Oct  3 22:10 .
drwxr-xr-x@ 11 jan  staff  352 Oct  3 22:09 ..
drwxr-xr-x@ 12 jan  staff  384 Oct  3 22:11 .git
lrwxr-xr-x@  1 jan  staff   19 Oct  3 22:10 .lima.yaml -> templates/demo.yaml
drwxr-xr-x@  3 jan  staff   96 Oct  3 22:09 templates

❯ l tmpl url github:jandubois
https://raw.githubusercontent.com/jandubois/jandubois/main/templates/demo.yaml
```

The symlink check is actually done for any file called `.lima.yaml` or `lima.yaml`. When we implement #4135 we can do it for all URLs and read the whole file, and just return the content if the file isn't a symlink.

@AkihiroSuda I don't think the tests should reference my personal repos; can I create `lima-vm/lima-vm` with the same content?

The tests should probably move to BATS later, but there is a lot of plugin functionality that still needs tests.

jandubois commented:
There is one other change I want to make: the branch/tag/sha should follow the repo, not the path/filename: `github:ORG/REPO@TAG/PATH` instead of `github:ORG/REPO/PATH@TAG`.

I think I originally put it last because it comes last in `go.mod` etc, but that is irrelevant for Lima templates. The branch/tag/sha is a property of the repo and not the file, so it should be closer to the repo name.

Additional benefits:

* The template filename remains at the end of the URL instead of being sandwiched.
* It mirrors the order in the `githubusercontent` replacement URL.

Not super important, but feels better to me.

jandubois commented:
> > Why both? The dot file seems better for the specially reserved file name. Is there any case the dot file has to be avoided?
> 
> So you can have `lima.yaml` files in other places in your repo that are symlinks. The non-dot version is because the default filename is `lima.yaml` if the URL ends with a slash. We could make it also `.lima.yaml`, but I think if the file is in a subdirectory, I would not want to have it hidden.

I changed my mind on this. It is simpler to say "when only a directory is specified, then the filename will be `.lima.yaml`". And we don't really want to see the `lima.yaml` file in repositories.

By using `.lima.yaml` everywhere, every directory can have a hidden symlink that is used when you specify just the directory (with a trailing slash), which seems more useful.

Thanks for questioning this; it was not a good idea to support `lima.yaml` as the default.

jandubois commented:
I've rewritten the tests using BATS:

```console
❯ ./lib/bats-core/bin/bats -T tests/url-github.bats
url-github.bats
 ✓ github:jandubois/jandubois/templates/demo.yaml@main [79]
 ✓ github:jandubois/jandubois/templates/demo.yaml [465]
 ✓ github:jandubois/jandubois/templates/demo [414]
 ✓ github:jandubois/jandubois/.lima.yaml [566]
 ✓ github:jandubois/jandubois/@v0.0.0 [206]
 ✓ github:jandubois/jandubois [414]
 ✓ github:jandubois//templates/demo.yaml@main [40]
 ✓ github:jandubois//templates/demo.yaml [358]
 ✓ github:jandubois//templates/demo [391]
 ✓ github:jandubois//.lima.yaml [419]
 ✓ github:jandubois//@v0.0.0 [112]
 ✓ github:jandubois// [457]
 ✓ github:jandubois/ [409]
 ✓ github:jandubois@v0.0.0 [103]
 ✓ github:jandubois [408]
 ✓ github:jandubois/jandubois/docs/.lima.yaml@main [222]
 ✓ github:jandubois/jandubois/docs/.lima.yaml [434]
 ✓ github:jandubois/jandubois/docs/.lima [652]
 ✓ github:jandubois/jandubois/docs/ [466]
 ✓ github:jandubois//docs/.lima.yaml@v0.0.0 [209]
 ✓ github:jandubois//docs/.lima.yaml [441]
 ✓ github:jandubois//docs/.lima [545]
 ✓ github:jandubois//docs/@v0.0.0 [109]
 ✓ github:jandubois//docs/ [428]
 ✓ .lima.yaml is retained when it is not a symlink [530]
 ✓ hidden files without an extension get a .yaml extension [398]
 ✓ files that have an extension do not get a .yaml extension [357]
 ✓ github: URLs are EXPERIMENTAL [445]
 ✓ Empty github: url returns an error [49]
 ✓ Missing org returns an error [47]

30 tests, 0 failures in 12 seconds
```

I've also simplified the rules if a file looks like a symlink:

Must be non-empty and have no newline, space, or colon. That's it.

jandubois commented:
> I'm somewhat surprised to see that the API call to determine the default branch adds 300-400ms to the processing time. The symlink resolution only takes an additional 50-100ms.

Out of curiosity I've sorted the individual tests into the 4 different buckets:

### No network requests

```
 ✓ github:jandubois/jandubois/templates/demo.yaml@main [79]
 ✓ github:jandubois//templates/demo.yaml@main [40]
 ✓ Empty github: url returns an error [49]
 ✓ Missing org returns an error [47]
```

### Symlink resolution via githubusercontent

```
 ✓ github:jandubois/jandubois/@v0.0.0 [206]
 ✓ github:jandubois//@v0.0.0 [112]
 ✓ github:jandubois@v0.0.0 [103]
 ✓ github:jandubois/jandubois/docs/.lima.yaml@main [222]
 ✓ github:jandubois//docs/.lima.yaml@v0.0.0 [209]
 ✓ github:jandubois//docs/@v0.0.0 [109]
```

### Default branch lookup via GitHub API

```
 ✓ github:jandubois/jandubois/templates/demo.yaml [465]
 ✓ github:jandubois/jandubois/templates/demo [414]
 ✓ github:jandubois//templates/demo.yaml [358]
 ✓ github:jandubois//templates/demo [391]
 ✓ hidden files without an extension get a .yaml extension [398]
 ✓ files that have an extension do not get a .yaml extension [357]
```

### Both symlink and branch requests

```
 ✓ github:jandubois/jandubois/.lima.yaml [566]
 ✓ github:jandubois/jandubois [414]
 ✓ github:jandubois//.lima.yaml [419]
 ✓ github:jandubois// [457]
 ✓ github:jandubois/ [409]
 ✓ github:jandubois [408]
 ✓ github:jandubois/jandubois/docs/.lima.yaml [434]
 ✓ github:jandubois/jandubois/docs/.lima [652]
 ✓ github:jandubois/jandubois/docs/ [466]
 ✓ github:jandubois//docs/.lima.yaml [441]
 ✓ github:jandubois//docs/.lima [545]
 ✓ github:jandubois//docs/ [428]
 ✓ .lima.yaml is retained when it is not a symlink [530]
 ✓ github: URLs are EXPERIMENTAL [445]
```

jandubois commented:
I got one more idea today:

We could check if the `.lima.yaml` "symlink" is actually an `http://` or `https://` URL, and then treat it as a redirect instead of a symlink.

This would allow you to have `github:ORG` as a short canonical URL, which points to `ORG/ORG/.lima.yaml`, which could then redirect to a template in a different repo.

That way you don't need to keep the template in the ORG/ORG repo, which may be unwanted.

The problem with this idea is that you cannot specify a tag in the template repo this way, it would always resolve to a branch/tag in the ORG/ORG repo.

I guess it would be up to the ORG maintainers to decide how they want to structure their repos, and which `github:` URLs they want to promote. Just because this feature exists doesn't mean you have to use it.

This can always be added later, so I think this PR should be merged as-is for now.

jandubois commented:
> This is probably fine, but we have to confirm whether we need some sort of "CORS" restriction; do we need to disallow a redirect to a different org?

IDK. It feels inconsistent because we don't restrict redirects on http/https template URLs to the same domain:

```shell
limactl create https://example.com/templates/example.yaml
```

This can redirect wherever it likes, and we would still load the template.

However, this feature is only meant to be used to redirect from the ORG repo to a different repo within the same ORG, so it feels fine to me to restrict this, at least initially. But this means also that we only allow `github:` URLs as the redirect targets, and not http/https ones. Which again feels fine, because all allowed locations are accessible by the `github:` scheme. And http/https doesn't allow the tag propagation.

It is always better to only allow a subset at first, and expand later based on feedback, instead of providing it all, and having to take some of it away again later.

Anyways, I hope this PR can be approved as-is, and this feature can be a separate PR.

AkihiroSuda reviewed (approved):
Thanks

We need a documentation, probably as a subpage of https://lima-vm.io/docs/templates/ 
(`https://lima-vm.io/docs/templates/github`)

jandubois commented on hack/bats/tests/url-github.bats:
```diff
@@ -0,0 +1,102 @@
+	github:jandubois//
+	github:jandubois/
+	github:jandubois@v0.0.0
+	github:jandubois
```
We could shorten it, but `template:` is kind of long too, so `github:` fits right in, and maybe `gh:` is a bit obscure for some people.

Or did you want `gh:` as an alias, and not as a replacement? Not sure if that is good, as people might wonder what the difference would be.

AkihiroSuda replied:
As an alias. No need to cover in this PR anyway.

alexandear commented on pkg/limatmpl/github.go:
Is it possible to add unit tests for functions in the file?

alexandear commented on pkg/limatmpl/github.go:
Maybe it would be better to use https://github.com/google/go-github for making requests to GitHub?

AkihiroSuda replied:
No need to inflate the dependency for this amount of code
